### ISSUE TITLE
This machine is sending different environment parameters than the first machine that started this parallel run.

### ISSUE THREAD

**Issue by sean-hill** [cypress-io/github-action#602]
Hey there! I'm receiving this error just recently in my Github Actions that run multiple tests in parallel:

```
/opt/hostedtoolcache/node/14.20.1/x64/bin/npx cypress cache list
┌─────────┬───────────────────┐
│ version │ last used         │
├─────────┼───────────────────┤
│ 10.3.0  │ a few seconds ago │
└─────────┴───────────────────┘
[2200:1004/193837.782206:ERROR:sandbox_linux.cc(377)] InitializeSandbox() called with multiple threads in process gpu-process.
[2200:1004/193837.793000:ERROR:gpu_memory_buffer_support_x11.cc(44)] dri3 extension not supported.
Missing baseUrl in compilerOptions. tsconfig-paths will be skipped
You passed the --parallel flag, but we do not parallelize tests across different environments.

This machine is sending different environment parameters than the first machine that started this parallel run.

The existing run is: https://dashboard.cypress.io/projects/<:)>/runs

In order to run in parallel mode each machine must send identical environment parameters such as:

 - specs
 - osName
 - osVersion
 - browserName
 - browserVersion (major)

This machine sent the following parameters:

{
  "osName": "linux",
  "osVersion": "Ubuntu - 20.04",
  "browserName": "Chrome",
  "browserVersion": "106.0.5249.91",
  "specs": [
    ...
  ]
}

https://on.cypress.io/parallel-group-params-mismatch
Test run failed, code 1
More information might be available above
Cypress module has returned the following error message:
Could not find Cypress test run results
Error: Could not find Cypress test run results
```

I can confirm that a view of the jobs are running Chrome 105 and other Chrome 106. Here is my workflow configuration:

```
- name: Run Cypress tests
  id: e2e-tests
  uses: cypress-io/github-action@v4
  with:
    browser: chrome
    record: true
    parallel: true
    group: 'Transact'
    config: baseUrl=${{ github.event.inputs.baseUrl }}
    env: apiHost=${{ github.event.inputs.apiHost }}
  env:
    CYPRESS_RECORD_KEY: ${{ secrets.TRANSACT_CYPRESS_RECORD_KEY }}
    GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
 ```

Is Github using different versions of chrome on their `ubuntu-latest` image? 🤔

**Comment by MaximeHeckel** [comment]
Same here, happens every few weeks 🤷‍♂️

**Comment by phishy** [comment]
I'm so done with Cypress.

**Comment by phishy** [comment]
I've pinned both the ubuntu versions and the actions version. Nada.

**Comment by openam** [comment]
I see different caches in the logs. I may need to try specifying a custom cache key https://github.com/cypress-io/github-action#custom-cache-key
```
/opt/hostedtoolcache/node/14.20.0/x64/bin/npx cypress cache list
```
vs 
```
/opt/hostedtoolcache/node/14.20.1/x64/bin/npx cypress cache list
```

**Comment by jonkoops** [comment]
This has started happening for us on our repo as well: https://github.com/keycloak/keycloak-ui

**Comment by jonkoops** [comment]
@openam `cache-key` has nothing to do with this I think. That is the key used to cache dependencies.

**Comment by nicolasvienot** [comment]
Hello! We are experiencing the same issue with parallel Cypress tests in our GitHub CI. It has already happened a few times in the past, but it was rare. Since the start of the week, it happens a lot and more than half of our test job fails.

**Comment by phishy** [comment]
@bahmutov Has informed me, that even when pinning Ubuntu versions, GitHub will upgrade the browser and Cypress will complain. The advice is that it is best to use a [cypress docker container](https://github.com/cypress-io/cypress-docker-images/tree/master/browsers) in order to pin the browser version.

**Comment by jonkoops** [comment]
Yeah, I figured that would be the reason why. I guess this is not a bug on the side if the action, but perhaps a better error message could be shown to point users to this solution.

**Comment by bahmutov** [comment]
One could also maybe control how exactly the browser versions are matched? I doubt that chrome update X.Y.Z.101 to X.Y.Z.105 would matter much

**Comment by jonkoops** [comment]
My main problem with the error message is that I'd like to see what the actual difference is between the environment parameters. It would make it a lot easier to identify the exact parameter that causes the issue.

**Comment by beamery-tomht** [comment]
Any updates on this? It happens for us almost monthly and seems to be whenever chrome updates!  I agree that the check against browser matching for environments could be configured to solve for this issue and be less strict

**Comment by orizens** [comment]
i'm trying to use https://github.com/marketplace/actions/setup-chrome to setup a chrome version - but seems like the `cypress-io/github-action@v4` action is picking up a different version.
is there anyway to set the chrome binary to cypress-io action?

**Comment by beamery-tomht** [comment]
Have got this working with setup-chrome after finding this comment https://github.com/cypress-io/github-action/issues/518#issuecomment-1210979047

As we use a custom command, had to specify `--browser ${{ env. BROWSER_PATH }}` in that command

**Comment by orizens** [comment]
thanks @beamery-tomht it works well!

**Comment by chinanderm** [comment]
Been happening all day today so far.

**Comment by MikeMcC399** [comment]
@chinanderm 

Thanks for your details! That ties in with the current [ubuntu-22.04](https://github.com/actions/runner-images/blob/main/images/linux/Ubuntu2204-Readme.md) rollout according to https://github.com/actions/runner-images

![image](https://user-images.githubusercontent.com/66998419/231527286-01c212ba-1d03-49d3-bca2-06744a5bf895.png)

Image version | Chrome version
--- | ---
20230409.1 | 112.0.5615.49
20230402.1 | 111.0.5563.146

so currently GitHub is 79% on Chrome `111` and 21% on Chrome `112`.

It seems the only workaround right now is to use a Docker container which ties to a specific version of Chrome.

This isn't something which can be solved by the `github-action`.

Perhaps there is scope here for a feature request to Cypress Cloud to allow adjacent major versions like `111` and `112` to be accepted? I don't know if that is technically feasible however. It seems like a reasonable request, since `runs-on:` is just specifying `ubuntu-latest` with no control about the exact Chrome version which GitHub is providing.

**Comment by nagash77** [comment]
This is not something that can be controlled by the github-action.  I am going to close this issue.

**Comment by MikeMcC399** [comment]
- Continued in https://github.com/cypress-io/github-action/issues/865

**Comment by MikeMcC399** [comment]
@chinanderm 

- In your workflow you are starting 4 containers and GitHub is using two different images for these containers, even though they are in the same job. This feels to me like it shouldn't happen and there should be consistency, despite any ongoing rollout. I've opened a discussion https://github.com/actions/runner-images/discussions/7436 to get feedback from GitHub.

```
    strategy:
      fail-fast: false
      matrix:
        # run copies of the current job in parallel
        containers: [1, 2, 3, 4]
```

If I were a productive and impacted user, I would consider submitting a concrete issue in https://github.com/actions/runner-images/issues, however I am only an observer, so I restricted myself to opening a discussion item (see above).

- I tried out the example [.github/workflows/example-custom-ci-build-id.yml](https://github.com/cypress-io/github-action/blob/master/.github/workflows/example-custom-ci-build-id.yml), adding `--browser chrome`, however I am consistently getting the old `ubuntu-22.04` image `Version: 20230402.1`, so I couldn't reproduce it. There is no transparency to the staged rollout apart from the percentage completion, so I don't know when I might be able to reproduce the issue during the rollout.

![image](https://user-images.githubusercontent.com/66998419/231718896-699d5e94-5802-4e10-bc80-42d2e71d6cdd.png)

**Comment by MikeMcC399** [comment]
@bahmutov 

> This is GitHub actions rolling new virtual machines with updates chrome browser. It takes time so some containers get version X, others X+1

Thanks for your confirmation about the root cause. Do you know if this has ever been discussed with GitHub? Are they able to stop it happening?

**Comment by MikeMcC399** [comment]
@bahmutov 
>Do you know if this has ever been discussed with GitHub? Are they able to stop it happening?
>> Hehehe they won’t stop rolling new versions. It is up to cypress to not freak out when chrome updates from 111 to 112 as if it is the end of the world 

I meant to ask if it had been discussed for GitHub not to use different image versions in the same job and same matrix, despite a rollout being in progress. I was not suggesting that GitHub stop rolling out new versions. I was just suggesting that they might control the image usage better during a rollout.

**Comment by bahmutov** [comment]
I don’t think it was ever discussed. Sent from my iPhoneOn Apr 13, 2023, at 07:44, Mike McCready ***@***.***> wrote:﻿
@bahmutov

Do you know if this has ever been discussed with GitHub? Are they able to stop it happening?

Hehehe they won’t stop rolling new versions. It is up to cypress to not freak out when chrome updates from 111 to 112 as if it is the end of the world 


I meant to ask if it had been discussed for GitHub not to use different image versions in the same job and same matrix, despite a rollout being in progress. I was not suggesting that GitHub stop rolling out new versions. I was just suggesting that they might control the image usage better during a rollout.

**Comment by sgronblo** [comment]
@MikeMcC399 Thanks for trying to bring up the issue with Github.

We are currently seeing failures due to this exact issue. Instead of asking everyone who uses --parallel with GH Actions to set up the Chrome version manually through additional actions or docker containers, could we not just have a simple setting for Cypress to specify how many major versions of drift we are ready to accept. Or a way to temporarily disable the check while the rollout is going on.

**Comment by sgronblo** [comment]
Am I understanding correctly from this that if I want to use one of Cypress's official Docker images here: https://github.com/cypress-io/cypress-docker-images/tree/master/browsers I am limited to Chrome 107 as the latest version?

**Comment by MikeMcC399** [comment]
@sgronblo 

> Am I understanding correctly from this that if I want to use one of Cypress's official Docker images here: https://github.com/cypress-io/cypress-docker-images/tree/master/browsers I am limited to Chrome 107 as the latest version?

The list you refer to is incomplete. See https://hub.docker.com/r/cypress/browsers/tags/ instead. (The current latest Chrome version available in the Docker images in this category is 113.)

- I reported this documentation issue in https://github.com/cypress-io/cypress-docker-images/issues/889

**Comment by MikeMcC399** [comment]
@sgronblo 

> We are currently seeing failures due to this exact issue. Instead of asking everyone who uses --parallel with GH Actions to set up the Chrome version manually through additional actions or docker containers, could we not just have a simple setting for Cypress to specify how many major versions of drift we are ready to accept. Or a way to temporarily disable the check while the rollout is going on.

- This is a reasonable suggestion. https://github.com/cypress-io/cypress/issues/25040 is the open issue on the Cypress side which you can look at and add any comments you like there. It is not something that `github-action` can do anything about. Any solution needs to come from GitHub and/or the core Cypress product.
